roachtest: sqlsmith field year value is out of range

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2433392&tab=buildLog) on [release-20.2@abeb6a501e00a23477aab67951a9b511b7ce7ee5](https://github.com/cockroachdb/cockroach/commits/abeb6a501e00a23477aab67951a9b511b7ce7ee5):

```
The test failed on branch=release-20.2, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:181,sqlsmith.go:211,test_runner.go:755: error: pq: internal error: '0000-12-31 00:00:00':::TIMESTAMP::TIMESTAMP: parsing as type timestamp: field year value 0 is out of range
		stmt:
		SELECT
			tab_4100._date AS col_9952,
			(
				SELECT
					('0001-01-01 00:00:00':::TIMESTAMP::TIMESTAMP - tab_4101._interval::INTERVAL)::TIMESTAMP AS col_9953
				FROM
					defaultdb.public.seed@[0] AS tab_4101
				LIMIT
					1:::INT8
			)
				AS col_9954,
			tab_4100._timestamptz AS col_9955,
			tab_4100._string AS col_9956,
			tab_4100._uuid AS col_9957,
			tab_4100._inet AS col_9958
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_4100
		LIMIT
			34:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2433392&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/datetime
+++ b/pkg/sql/logictest/testdata/logic_test/datetime
@@ -113,8 +113,8 @@ SELECT 'epoch'::date, 'infinity'::date, '-infinity'::date
 
 # Date edge cases.
 
-statement error year value 0 is out of range
-SELECT '0000-01-01'::date
+statement error only positive years are permitted in AD/BC notation
+SELECT '0000-01-01 BC'::date
 
 query TTTTT
 SELECT '4714-11-24 BC'::date, '5874897-12-31'::date, '2000-01-01'::date, '0001-01-01'::date, '0001-12-13 BC'::date

--- a/pkg/sql/logictest/testdata/logic_test/distsql_datetime
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_datetime
@@ -1,0 +1,12 @@
+# LogicTest: 5node-default-configs
+
+statement ok
+CREATE TABLE ts (a INT PRIMARY KEY, t TIMESTAMP);
+INSERT INTO ts SELECT i, '0001-01-01 00:00:00'::TIMESTAMP + i::INTERVAL FROM generate_series(1, 5) AS g(i);
+ALTER TABLE ts SPLIT AT select i FROM generate_series(2, 5) AS g(i);
+ALTER TABLE ts EXPERIMENTAL_RELOCATE SELECT ARRAY[i%5+1], i FROM generate_series(1, 5) AS g(i)
+
+# This query makes sure that we can successfully deserialize the timestamp with
+# year 0 on the remote nodes (#56554).
+statement ok
+SELECT t - (SELECT '0001-01-01 00:00:00'::TIMESTAMP - a::INTERVAL FROM ts ORDER BY a LIMIT 1) FROM ts

--- a/pkg/util/timeutil/pgdate/field_extract.go
+++ b/pkg/util/timeutil/pgdate/field_extract.go
@@ -884,18 +884,20 @@ func (fe *fieldExtract) validate() error {
 	}
 
 	if year, ok := fe.Get(fieldYear); ok {
-		if year == 0 {
-			return outOfRangeError("year", year)
-		}
-		// Update for BC dates.
-		if era, ok := fe.Get(fieldEra); ok && era < 0 {
-			// No year 0
+		// Note that here we allow for year to be 0 (which means 1 BC) which is
+		// a deviation from Postgres. The issue is that we support two notations
+		// (numbers or numbers with AD/BC suffix) whereas Postgres supports only
+		// the latter.
+
+		if era, ok := fe.Get(fieldEra); ok {
 			if year <= 0 {
-				return inputErrorf("no year 0 in AD/BC notation")
+				return inputErrorf("only positive years are permitted in AD/BC notation")
 			}
-			// Normalize to a negative year
-			if err := fe.Reset(fieldYear, 1-year); err != nil {
-				return err
+			if era < 0 {
+				// Update for BC dates.
+				if err := fe.Reset(fieldYear, 1-year); err != nil {
+					return err
+				}
 			}
 		} else if fe.tweakYear {
 			if year < 0 {

--- a/pkg/util/timeutil/pgdate/parsing_test.go
+++ b/pkg/util/timeutil/pgdate/parsing_test.go
@@ -397,6 +397,14 @@ var dateTestData = []timeData{
 		s:   "121212",
 		exp: time.Date(2012, 12, 12, 0, 0, 0, 0, time.UTC),
 	},
+	{
+		s:   "-0001-02-15",
+		exp: time.Date(-1, 2, 15, 0, 0, 0, 0, time.UTC),
+	},
+	{
+		s:   "0000-02-15",
+		exp: time.Date(0, 2, 15, 0, 0, 0, 0, time.UTC),
+	},
 }
 
 var timeTestData = []timeData{

--- a/pkg/util/timeutil/pgdate/pgdate_test.go
+++ b/pkg/util/timeutil/pgdate/pgdate_test.go
@@ -65,8 +65,8 @@ func TestParseDate(t *testing.T) {
 			err: "date is out of range",
 		},
 		{
-			s:   "0000-01-01",
-			err: "year value 0 is out of range",
+			s:   "0000-01-01 AD",
+			err: "only positive years are permitted in AD/BC notation",
 		},
 	} {
 		t.Run(tc.s, func(t *testing.T) {

